### ISSUE TITLE
compile for debian 10 implicit declaration of function 'readv'

### ISSUE THREAD

**Issue by geosone** [Castaglia/proftpd-mod_proxy_protocol#13]
```
dh build --without python-support
   dh_update_autotools_config
   debian/rules override_dh_auto_build
make[1]: Entering directory '/build/proftpd-mod-proxy-protocol-0.1'
DESTDIR=/build/proftpd-mod-proxy-protocol-0.1/debian/proftpd-mod-proxy-protocol CFLAGS=-Wno-implicit-function-declaration prxs -c mod_proxy_protocol.c
libtool --mode=compile gcc  -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-gtGFdY/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -c mod_proxy_protocol.c
libtool: compile:  gcc -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-gtGFdY/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -c mod_proxy_protocol.c  -fPIC -DPIC -o .libs/mod_proxy_protocol.o
mod_proxy_protocol.c: In function 'readv_sock':
mod_proxy_protocol.c:200:9: error: implicit declaration of function 'readv'; did you mean 'read'? [-Werror=implicit-function-declaration]
   res = readv(sockfd, iov, count);
         ^~~~~
         read
cc1: some warnings being treated as errors
prxs: error executing command (1)`
```
could you fix this its an unofficial debain package i am maintaining.
i trird to work around it but it would require a complete build system for the module.

http://git.disconnected-by-peer.at/websms/proftpd-mod-proxy_protocol/

**Comment by Castaglia** [comment]
@geosone What platform are you using for this build?  Do you happen to have a Dockerfile, or a Vagrant box, that I could use to reproduce the build issue?

**Comment by geosone** [comment]
its compiled in an cow image (its just debian10) https://pastebin.com/Z34wiqdv
its an internal build system i cant give you access 
its just plain debian 10 latest.

proftpd-basic_1.3.6-4+deb10u3_amd64.deb

**Comment by Castaglia** [comment]
Hmm.  Does that platform have the `libc6-dev` package installed?  That provides the system headers (such as `/usr/include/linux/uio.h`)?  If so, it _could_ be that the `readv(2)` function, which _should_ be declared in `<sys/uio.h>`, has been _moved_ to `<linux/uio.h>` for some reason.  Which, per the man page, it should *not* have been.

Or maybe it's something else?  I installed the `libc6-dev` package on a `debian:10` Docker container I'm running, then looked through the system headers for where `readv` is declared:
```
/usr/include/x86_64-linux-gnu/sys/uio.h:			   preadv64) __wur;
/usr/include/x86_64-linux-gnu/sys/uio.h:#   define preadv preadv64
/usr/include/x86_64-linux-gnu/sys/uio.h:extern ssize_t preadv64 (int __fd, const struct iovec *__iovec, int __count,
```

Only _after_ installing `libc6-dev` did I run ProFTPD's configure script:
```
$ apt-get install -y gcc libc6-dev make
$ cd /src/proftpd
$ ./configure
```
Specifically, I want to see if the configure script picked up the `<sys/uio.h>` header; that's the header file which declares the `readv(2)` function.
```
root@02adf3393344:/src/proftpd# grep UIO config.h 
#define HAVE_SYS_UIO_H 1
```
That means the Autoconf script _did_ see the `<sys/uio.h>` header, as it should.

**Comment by geosone** [comment]
libc6-dev is a basic requirement of build-essetials
https://packages.debian.org/de/buster/build-essential

in short
apt-get install build-essential
apt-get install proftpd-dev
prxs -c mod_proxy_protocol.c

is happaning in the build system. in short

**Comment by Castaglia** [comment]
Hrm:
```
root@02adf3393344:/src/proftpd# apt-get install -y proftpd-dev
root@02adf3393344:/src/proftpd# grep -nr UIO /usr/include/proftpd/
/usr/include/proftpd/mod_sftp.h:38:#if HAVE_SYS_UIO_H
/usr/include/proftpd/mod_snmp.h:53:#if HAVE_SYS_UIO_H
/usr/include/proftpd/config.h:836:#define HAVE_SYS_UIO_H 1
```
That looks correct.

So then I did:
```
root@02adf3393344:/src/proftpd# git clone https://github.com/Castaglia/proftpd-mod_proxy_protocol.git /tmp/mod_proxy_protocol   
...
root@02adf3393344:/src/proftpd# cd /tmp/mod_proxy_protocol/
root@02adf3393344:/tmp/mod_proxy_protocol# prxs -c mod_proxy_protocol.c
libtool --mode=compile gcc  -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-h285Mw/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -c mod_proxy_protocol.c
libtool: compile:  gcc -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-h285Mw/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -c mod_proxy_protocol.c  -fPIC -DPIC -o .libs/mod_proxy_protocol.o
libtool: compile:  gcc -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-h285Mw/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -c mod_proxy_protocol.c -o mod_proxy_protocol.o >/dev/null 2>&1
libtool --mode=link gcc -o mod_proxy_protocol.la -rpath /usr/lib/proftpd  -avoid-version -export-dynamic -module  mod_proxy_protocol.lo 
libtool: link: gcc -shared  -fPIC -DPIC  .libs/mod_proxy_protocol.o      -Wl,-soname -Wl,mod_proxy_protocol.so -o .libs/mod_proxy_protocol.so
libtool: link: ar cru .libs/mod_proxy_protocol.a  mod_proxy_protocol.o
ar: `u' modifier ignored since `D' is the default (see `U')
libtool: link: ranlib .libs/mod_proxy_protocol.a
libtool: link: ( cd ".libs" && rm -f "mod_proxy_protocol.la" && ln -s "../mod_proxy_protocol.la" "mod_proxy_protocol.la" )
```

So then, let's use the `CFLAGS` you mention:
```
root@02adf3393344:/tmp/mod_proxy_protocol# CFLAGS=-Wno-implicit-function-declaration prxs -c mod_proxy_protocol.c
libtool --mode=compile gcc  -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-h285Mw/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -c mod_proxy_protocol.c
libtool: compile:  gcc -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-h285Mw/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -c mod_proxy_protocol.c  -fPIC -DPIC -o .libs/mod_proxy_protocol.o
libtool: compile:  gcc -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-h285Mw/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -c mod_proxy_protocol.c -o mod_proxy_protocol.o >/dev/null 2>&1
libtool --mode=link gcc -o mod_proxy_protocol.la -rpath /usr/lib/proftpd  -avoid-version -export-dynamic -module  mod_proxy_protocol.lo 
libtool: link: rm -fr  .libs/mod_proxy_protocol.a .libs/mod_proxy_protocol.la .libs/mod_proxy_protocol.lai .libs/mod_proxy_protocol.so
libtool: link: gcc -shared  -fPIC -DPIC  .libs/mod_proxy_protocol.o      -Wl,-soname -Wl,mod_proxy_protocol.so -o .libs/mod_proxy_protocol.so
libtool: link: ar cru .libs/mod_proxy_protocol.a  mod_proxy_protocol.o
ar: `u' modifier ignored since `D' is the default (see `U')
libtool: link: ranlib .libs/mod_proxy_protocol.a
libtool: link: ( cd ".libs" && rm -f "mod_proxy_protocol.la" && ln -s "../mod_proxy_protocol.la" "mod_proxy_protocol.la" )
```
Hmm.  Still succeeds?

**Comment by geosone** [comment]
ok so aktual master works but 0.1 dont ok

**Comment by Castaglia** [comment]
Ah, you mean the commit/fix for Issue #6 ...

**Comment by Castaglia** [comment]
I've just published an 0.2 release: https://github.com/Castaglia/proftpd-mod_proxy_protocol/releases/tag/v0.2

**Comment by geosone** [comment]
the newest 0.2 releas comples without problem thx.
`deb https://debdev.disconnected-by-peer.at/proftpd-mod-proxy-protocol-buster buster main`
also for stretch jessie wheezy-lts just adopt the line
